File: app/src/main/java/com/codemybrainsout/imageviewer/view/user/UserActivity.java
package com.codemybrainsout.imageviewer.view.user;

import android.animation.Animator;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.CoordinatorLayout;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.view.MenuItem;
import android.view.ViewAnimationUtils;
import android.view.animation.AccelerateInterpolator;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.codemybrainsout.imageviewer.R;
import com.codemybrainsout.imageviewer.api.UserService;
import com.codemybrainsout.imageviewer.custom.CircleImageView;
import com.codemybrainsout.imageviewer.custom.SquareImageView;
import com.codemybrainsout.imageviewer.listener.EndlessRecyclerViewScrollListener;
import com.codemybrainsout.imageviewer.listener.RecyclerViewItemClickListener;
import com.codemybrainsout.imageviewer.model.BaseModel;
import com.codemybrainsout.imageviewer.model.Photo;
import com.codemybrainsout.imageviewer.model.User;
import com.codemybrainsout.imageviewer.transformation.BlurTransformation;
import com.codemybrainsout.imageviewer.view.base.BaseActivity;
import com.codemybrainsout.imageviewer.view.detail.DetailActivity;
import com.codemybrainsout.imageviewer.view.home.MultiViewAdapter;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;

public class UserActivity extends BaseActivity implements UserContract.View, RecyclerViewItemClickListener {

    private static final String EXTRA_USER = "extra_user";

    @BindView(R.id.activity_user_toolbar)
    Toolbar activityUserToolbar;
    @BindView(R.id.activity_user_background_IV)
    ImageView activityUserBackgroundIV;
    @BindView(R.id.activity_user_profile_IV)
    CircleImageView activityUserProfileIV;
    @BindView(R.id.activity_user_name_TV)
    TextView activityUserNameTV;
    @BindView(R.id.activity_users_followers_TV)
    TextView activityUsersFollowersTV;
    @BindView(R.id.activity_users_following_TV)
    TextView activityUsersFollowingTV;
    @BindView(R.id.activity_users_social_LL)
    LinearLayout activityUsersSocialLL;
    @BindView(R.id.activity_user_bio_TV)
    TextView activityUserBioTV;
    @BindView(R.id.activity_user_collapsing_toolbar)
    CollapsingToolbarLayout activityUserCollapsingToolbar;
    @BindView(R.id.activity_user_app_bar_layout)
    AppBarLayout activityUserAppBarLayout;
    @BindView(R.id.activity_user_photos_RV)
    RecyclerView activityUserPhotosRV;
    @BindView(R.id.parent)
    CoordinatorLayout parent;

    @Inject
    public MultiViewAdapter multiViewAdapter;

    @Inject
    public UserPresenter userPresenter;
    private int page = 1, limit = 20;
    private String username;

    public static Intent getUserActivityIntent(Context context, User user) {
        Intent intent = new Intent(context, UserActivity.class);
        intent.putExtra(EXTRA_USER, user);
        return intent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_user);
        ButterKnife.bind(this);
        setToolbar();

        UserComponent userComponent = DaggerUserComponent.builder()
                .appComponent(getAppComponent())
                .userModule(new UserModule(this))
                .build();
        userComponent.inject(this);

        multiViewAdapter.setRecyclerViewItemClickListener(this);
        StaggeredGridLayoutManager staggeredGridLayoutManager = new StaggeredGridLayoutManager(2, 1);
        activityUserPhotosRV.setLayoutManager(staggeredGridLayoutManager);
        activityUserPhotosRV.setAdapter(multiViewAdapter);
        activityUserPhotosRV.addOnScrollListener(new EndlessRecyclerViewScrollListener(staggeredGridLayoutManager) {
            @Override
            public void onLoadMore(int page, int totalItemsCount, RecyclerView view) {
                userPresenter.loadMorePhotos(username, page, limit);
            }
        });

        if (getIntent().hasExtra(EXTRA_USER)) {
            User user = (User) getIntent().getSerializableExtra(EXTRA_USER);
            username = user.getUsername();
            userPresenter.loadUser(username);
        }
    }

    private void beginEnterTransition() {
        activityUserBackgroundIV.post(new Runnable() {
            @Override
            public void run() {

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    final Animator circularReveal = ViewAnimationUtils.createCircularReveal(activityUserBackgroundIV, activityUserBackgroundIV.getWidth() / 2, 0, 0, activityUserBackgroundIV.getWidth());
                    circularReveal.setInterpolator(new AccelerateInterpolator(1.5f));
                    circularReveal.setDuration(500);
                    circularReveal.start();
                }

            }
        });
    }

    @Override
    public void showLoading() {
        addProgressLayout(this, parent);
    }

    @Override
    public void hideLoading() {
        removeProgressLayout(parent);
    }

    @Override
    public void showUser(User user) {

        if (!TextUtils.isEmpty(user.getProfileImage().getLarge())) {
            Glide.with(this)
                    .load(user.getProfileImage().getLarge())
                    .into(activityUserProfileIV);
        }

        if (!TextUtils.isEmpty(user.getProfileImage().getLarge())) {
            Glide.with(this)
                    .load(user.getProfileImage().getLarge())
                    .bitmapTransform(new BlurTransformation(this, 15))
                    .into(activityUserBackgroundIV);
        }

        beginEnterTransition();

        activityUserCollapsingToolbar.setTitleEnabled(true);
        activityUserCollapsingToolbar.setExpandedTitleColor(Color.TRANSPARENT);
        activityUserCollapsingToolbar.setCollapsedTitleTextColor(Color.WHITE);
        activityUserCollapsingToolbar.setTitle(TextUtils.isEmpty(user.getName()) ? "" : user.getName());

        activityUserNameTV.setText(TextUtils.isEmpty(user.getName()) ? "" : user.getName());
        activityUserBioTV.setText(TextUtils.isEmpty(user.getBio()) ? "" : user.getBio());
        activityUsersFollowersTV.setText(user.getFollowersCount() + " Followers");
        activityUsersFollowingTV.setText(user.getFollowingCount() + " Following");

        userPresenter.loadPhotos(user.getUsername(), page, limit);
    }

    @Override
    public void refreshPhotos(List<Photo> list) {
        multiViewAdapter.setPhotos(list);
    }

    @Override
    public void addPhotos(List<Photo> list) {
        multiViewAdapter.addPhotos(list);
    }

    @Override
    public void showPhotoDetails(Photo photo) {
        Intent intent = DetailActivity.getDetailActivityIntent(this, photo);
        startActivity(intent);
    }

    @Override
    public void showError(String s) {
        showErrorSnackbar(parent, s);
    }

    @Override
    public void onItemClick(BaseModel baseModel) {
        userPresenter.openPhotoDetails((Photo) baseModel);
    }

    @Override
    public void onUserClick(User user) {

    }

    @Override
    public void setToolbar() {

        setSupportActionBar(activityUserToolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayShowTitleEnabled(false);
            actionBar.setDisplayHomeAsUpEnabled(true);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        int id = item.getItemId();
        if (id == android.R.id.home) {
            onBackPressed();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

}

File: app/src/main/java/com/codemybrainsout/imageviewer/view/user/UserActivity.kt
package com.codemybrainsout.imageviewer.view.user

import android.animation.Animator
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.text.TextUtils
import android.view.MenuItem
import android.view.ViewAnimationUtils
import android.view.animation.AccelerateInterpolator
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.bumptech.glide.Glide
import com.codemybrainsout.imageviewer.R
import com.codemybrainsout.imageviewer.databinding.ActivityUserBinding
import com.codemybrainsout.imageviewer.extensions.addProgressLayout
import com.codemybrainsout.imageviewer.extensions.removeProgressLayout
import com.codemybrainsout.imageviewer.extensions.showErrorSnackbar
import com.codemybrainsout.imageviewer.extensions.viewBinding
import com.codemybrainsout.imageviewer.listener.EndlessRecyclerViewScrollListener
import com.codemybrainsout.imageviewer.listener.RecyclerViewItemClickListener
import com.codemybrainsout.imageviewer.model.BaseModel
import com.codemybrainsout.imageviewer.model.Photo
import com.codemybrainsout.imageviewer.model.User
import com.codemybrainsout.imageviewer.transformation.BlurTransformation
import com.codemybrainsout.imageviewer.view.detail.DetailActivity
import com.codemybrainsout.imageviewer.view.home.MultiViewAdapter
import javax.inject.Inject

class UserActivity : AppCompatActivity(), UserContract.View, RecyclerViewItemClickListener {

    private val binding by viewBinding(ActivityUserBinding::inflate)

    @Inject
    lateinit var userPresenter: UserPresenter

    private val page = 1
    private val limit = 20

    private val multiViewAdapter: MultiViewAdapter by lazy { MultiViewAdapter(this) }
    private var username: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding.apply {
            multiViewAdapter.setRecyclerViewItemClickListener(this@UserActivity)
            val staggeredGridLayoutManager = StaggeredGridLayoutManager(2, 1)
            recyclerViewPhotos.layoutManager = staggeredGridLayoutManager
            recyclerViewPhotos.adapter = multiViewAdapter
            recyclerViewPhotos.addOnScrollListener(object :
                EndlessRecyclerViewScrollListener(staggeredGridLayoutManager) {
                override fun onLoadMore(page: Int, totalItemsCount: Int, view: RecyclerView?) {
                    userPresenter.loadMorePhotos(username, page, limit)
                }
            })
        }

        if (intent.hasExtra(EXTRA_USER)) {
            val user = intent.getSerializableExtra(EXTRA_USER) as User
            username = user.username
            userPresenter.loadUser(username)
        }
    }

    private fun beginEnterTransition() {
        binding.imageViewBackground.post {
            val circularReveal: Animator = ViewAnimationUtils.createCircularReveal(
                binding.imageViewBackground,
                binding.imageViewBackground.width / 2,
                0,
                0f,
                binding.imageViewBackground.width.toFloat()
            )
            circularReveal.interpolator = AccelerateInterpolator(1.5f)
            circularReveal.duration = 500
            circularReveal.start()
        }
    }

    override fun showLoading() {
        binding.root.addProgressLayout(this)
    }

    override fun hideLoading() {
        binding.root.removeProgressLayout()
    }

    override fun showUser(user: User) {
        if (!TextUtils.isEmpty(user.profileImage!!.large)) {
            Glide.with(this)
                .load(user.profileImage!!.large)
                .into(binding.imageViewUser)
        }
        if (!TextUtils.isEmpty(user.profileImage!!.large)) {
            Glide.with(this)
                .load(user.profileImage!!.large)
                .transform(BlurTransformation(this, 15))
                .into(binding.imageViewBackground)
        }
        beginEnterTransition()
        binding.collapsingToolbar.isTitleEnabled = true
        binding.collapsingToolbar.setExpandedTitleColor(Color.TRANSPARENT)
        binding.collapsingToolbar.setCollapsedTitleTextColor(Color.WHITE)
        binding.collapsingToolbar.title = if (TextUtils.isEmpty(user.name)) "" else user.name
        binding.textViewUserName.text = if (TextUtils.isEmpty(user.name)) "" else user.name
        binding.textViewBio.text = if (TextUtils.isEmpty(user.bio)) "" else user.bio
        binding.textViewFollowers.text = "${user.followersCount} ${getString(R.string.followers)}"
        binding.textViewFollowing.text = "${user.followingCount} ${getString(R.string.following)}"
        userPresenter.loadPhotos(user.username, page, limit)
    }

    override fun refreshPhotos(list: List<Photo>) {
        multiViewAdapter.setPhotos(list)
    }

    override fun addPhotos(list: List<Photo>) {
        multiViewAdapter.addPhotos(list)
    }

    override fun showPhotoDetails(photo: Photo?) {
        val intent: Intent = DetailActivity.getDetailActivityIntent(this, photo)
        startActivity(intent)
    }

    override fun showError(s: String?) {
        s?.let { binding.root.showErrorSnackbar(it) }
    }

    override fun onItemClick(baseModel: BaseModel?) {
        userPresenter.openPhotoDetails(baseModel as Photo?)
    }

    override fun onUserClick(user: User?) {}

    fun setToolbar() {
        setSupportActionBar(binding.toolbar)
        supportActionBar?.let {
            it.setDisplayShowTitleEnabled(false)
            it.setDisplayHomeAsUpEnabled(true)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == android.R.id.home) {
            onBackPressed()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    companion object {
        private const val EXTRA_USER = "extra_user"
        fun getUserActivityIntent(context: Context?, user: User?): Intent {
            val intent = Intent(context, UserActivity::class.java)
            intent.putExtra(EXTRA_USER, user)
            return intent
        }
    }
}